spring-integration-ip/src/main/java/org/springframework/integration/ip/udp/DatagramPacketMessageMapper.java

```diff
@@ -142,7 +145,23 @@ public DatagramPacket fromMessage(Message<?> message) throws Exception {
 			buffer.put(bytes);
 			bytes = buffer.array();
 		}
-		return new DatagramPacket(bytes, bytes.length);
+		return getDatagramFromHeader(message, bytes, bytes.length);
+	}
+
+	private DatagramPacket getDatagramFromHeader(Message<?> message, byte[] bytes,
+												 int length) throws UnknownHostException {
+		InetAddress inetAddress = null;
+		if (message.getHeaders().get(IpHeaders.IP_ADDRESS, String.class) != null) {
+			inetAddress = InetAddress.getByName(message.getHeaders()
+					.get(IpHeaders.IP_ADDRESS, String.class));
```

Comment:
The original idea behind using external socket was that we have a server app builded with spring-integration and some "whatever" clients behind NAT. Inbound adapter is receiving UDP datagrams from clients and outbound adapter has to send responses to these clients through inbounds adapter socket (when other socket will be used NAT will block the traffic to clients).
And with this configuration in mind (outbound uses inbounds socket) headers are filled with clients host and port by inbound adapter. And when using socket expression we have to use these headers to know where to send responses.
But in other scenario we may want to use outbound adapter to send datagram to some arbitrary host & port.
And this is why we need to be able to use both headers and standalone parameters host/port for outbound adapter.